pkg/expression/util.go

```diff
@@ -494,7 +494,10 @@ func ColumnSubstituteImpl(ctx sessionctx.Context, expr Expression, schema *Schem
 					logutil.BgLogger().Error("Unexpected error happened during ColumnSubstitution", zap.Stack("stack"))
 					return false, failed, v
 				}
-				if oldCollEt.Collation == newCollEt.Collation {
+				// When column collations are consistent, constant propagation can proceed without checking the collation of the constants.
+				if len(v.GetArgs()) == 2 && v.GetArgs()[0].GetType().GetCollate() == v.GetArgs()[1].GetType().GetCollate() {
+					changed = true
+				} else if oldCollEt.Collation == newCollEt.Collation {
```

Comment:
<img width="785" alt="image" src="https://github.com/pingcap/tidb/assets/15825830/bcdf24f2-83d6-40e3-9ccf-443c64a00390">
from [link](https://dev.mysql.com/doc/refman/8.0/en/charset-connection.html) here, collation_connection does affect constant value's collation, while mysql plan indeed gives a quick return for one side of join is empty table (similar as quick point get and found nothing)

```
mysql> explain  SELECT  t.id1, t.id2, t.id3 FROM t1 t JOIN t2 ON t.id1 = t2.id1 WHERE t2.id1 = '63812828';
+----+-------------+-------+------------+------+---------------+------+---------+------+------+----------+--------------------------------+
| id | select_type | table | partitions | type | possible_keys | key  | key_len | ref  | rows | filtered | Extra                          |
+----+-------------+-------+------------+------+---------------+------+---------+------+------+----------+--------------------------------+
|  1 | SIMPLE      | NULL  | NULL       | NULL | NULL          | NULL | NULL    | NULL | NULL |     NULL | no matching row in const table |
+----+-------------+-------+------------+------+---------------+------+---------+------+------+----------+--------------------------------+
1 row in set, 1 warning (0.03 sec)
```
my advices is that you can dive a liiter bit of mysql opt code, figure out how them handle column's collation itself and those collation from constant value inherited from collation_connection